vault/diagnose/tls_verification_test.go

```diff
@@ -0,0 +1,317 @@
+	}
+	if !strings.Contains(err.Error(), "could not decode cert") {
+		t.Errorf("Bad error message: %w", err)
+	}
+}
+
+// TestTLSTrailingData uses a certificate from:
+// https://github.com/golang/go/issues/40545 that contains
+// an extra DER sequence, and makes sure a trailing data error
+// is returned.
+func TestTLSTrailingData(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./test-fixtures/trailingdatacert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSMinVersion:                 "tls10",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Error("TLS Config check on fake certificate should fail")
+	}
+	if !strings.Contains(err.Error(), "asn1: syntax error: trailing data") {
+		t.Errorf("Bad error message: %w", err)
+	}
+}
+
+// TestTLSExpiredCert checks that an expired certificate fails TLS checks
+// with an appropriate error.
+func TestTLSExpiredCert(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./test-fixtures/expiredcert.pem",
+				TLSKeyFile:                    "./test-fixtures/expiredprivatekey.pem",
+				TLSMinVersion:                 "tls10",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Error("TLS Config check on fake certificate should fail")
+	}
+	if !strings.Contains(err.Error(), "certificate has expired or is not yet valid") {
+		t.Errorf("Bad error message: %w", err)
+	}
+}
+
+// TestTLSMismatchedCryptographicInfo verifies that a cert and key of differing cryptographic
+// types, when specified together, is met with a unique error message.
+func TestTLSMismatchedCryptographicInfo(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./../../api/test-fixtures/keys/cert.pem",
+				TLSKeyFile:                    "./test-fixtures/ecdsa.key",
+				TLSMinVersion:                 "tls10",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Error("TLS Config check on fake certificate should fail")
+	}
+	if err.Error() != "tls: private key type does not match public key type" {
```

Comment:
did you want these to be `!string.Contains()` as you've done for your other tests?